contributors/design-proposals/scheduling/pod-preemption.md

```diff
@@ -106,59 +107,295 @@ Now, assume everything in the above example, but the best effort pod has priorit
+node of pod C to Node 1. C is moved to unschedulable list.
+*   Scheduler looks at pod D in the queue. Pod D can fit on Node 2.
+*   Scheduler goes ahead and binds pod D to Node 2 while pods A and B are in
+their graceful termination period and pod C is not bound yet.
 
-### Example 4
+##### Example 4
 
 ![image](images/preemption_4.png)
 
--  Everything is similar to example 1, but while scheduler is waiting for pods A and B to gracefully terminate, a new higher priority pod F is created and goes to the head of the queue.
--  Scheduler evaluates the feasibility of pod F and determines that it can be scheduled on Node 1. So, it sets the future node of pod F to Node 1.
--  Scheduler looks at pod C and assumes that pod F is already bound to Node 1, so pod C can no longer run on Node 1. Scheduler clears the future node of pod C.
--  Eventually when pods A and B terminate, pod F is bound to Node 1 and pods C and D remain pending.
+*   Everything is similar to example 1, but while scheduler is waiting for pods
+A and B to gracefully terminate, a new higher priority pod F is created and goes
+to the head of the queue.
+*   Scheduler evaluates the feasibility of pod F and determines that it can be
+scheduled on Node 1. So, it sets the nominated node name of pod F to Node 1 and
+places it in unschedulable list.
+*   Scheduler clears nominated node name of C and moves it to the scheduling queue.
+*   C is evaluated for scheduling, but it cannot be scheduled as pod F's nominated
+node name is set to Node 1.
+*   When B terminates, scheduler brings F, C, and D back to the scheduling queue.
+F is evaluated first. They cannot be scheduled.
+*   Eventually when pods A and B terminate, pod F is bound to Node 1 and pods C
+and D remain unschedulable.
+
+
+## Supporting PodDisruptionBudget
+
+Scheduler preemption will support PDB for Beta, but respecting PDB is not
+guaranteed. Preemption will try to avoid violating PDB, but if it doesn't find
+any lower priority pod to preempt without violating PDB, it goes ahead and
+preempts victims despite violating PDB. This is to guarantee that higher priority
+pods will always get precedence over lower priority pods in obtaining cluster resources.
+
+Here is what preemption will do:
+
+1.  When choosing victims on any evaluated node, preemption logic will try to
+reprieve pods whose PDBs are violated first. (In the alpha version, pods are
+reprieved by their ascending priority and PDB is ignored.)
+1.  In scoring nodes and choosing one for preemption, the number of pods whose
+PDBs are violated will be the most significant metric. So, a node with the lowest
+number of victims whose PDBs are violated is the one chosen for preemption. In
+the alpha version, most significant metric is the highest priority of victims.
+If there are more than one node with the same smallest number of victims whose
+PDBs are violated, lowest high priority victim will be used (as in alpha) and
+the rest of the metrics remain the same as before.
```

Comment:
Isn't that going to cause a sort of cascade-preemption? For example:
1. Pods A (priority=10, no-pdb) and pod B(priority=1, pdb preventing deletion) are running on a single node in cluster.
1. Pod C with higher priority shows up, but there is no more space in cluster.
1. Pod A is evicted to make place for C.
1. Pod C is scheduled.
1. Pod B is evicted to make space for A.

In the end PDB was violated anyway when preempting for pod A and pod A has been unnecessarily restarted. This example is, of course, artificial, but aren't we worried about the general pattern of evicting higher priority pods, only to see them preempt lower priority pods in a few seconds?

Reply:
The case you mentioned could happen. I can think of three possible solutions for it:
1. Let it happen. It causes a bit of extra preemption, but at the end it gets to the desired state.
2. Detect such a scenario and preempt the pod that will eventually be preempted in the first place. Detecting this case is not trivial and is computationally expensive, especially when there is a long chain of such pods.
3. Do not respect PDB in preemption. I don't think this is a good idea.

Among the three options, I think the first one is probably the best one. That's what is already proposed.